# deNOVOEnrich
Identification of _de novo_ transposon insertion sites from targeted sequencing data generated through Transposon display Sequencing (TEd-Seq)

<img src="images/denonvoEnrich_logo.jpg" alt="denovoEnrich logo" width="300" height="300"/>

## Overview

deNOVOEnrich is a computational pipeline developed to efficiently detect and profile genome-wide somatic transposition insertions, as well as non-reference heritable transposition events from targeted sequencing data generated through Transposon display sequencing (TEd-Seq). 

The pipeline leverages high-confidence split read alignments to identify TE:genome break sites, enabling accurate detection of bona-fide non-reference somatic and heritable new insertions of a transposon. Stringent filters were applied at several critical steps of the pipeline to ensure removal of false positives and reduction of background noise signal.

#### How does it work?
deNOVOEnrich processes raw paired-end reads sequentially to ultimately pinpoint the genomic location of novel transposon insertion sites at base resolution. The primary stages of the pipeline have been outlined below:

 - #### Retrieval of TE-specific reads from raw sequencing reads 
To enrich for reads of interest, the pipeline uses a short extremity sequence (~40-60bp, typically from the long terminal repeat or LTR) of the targeted transposable element (TE) family as a bait. This targeted screening isolates TE-specific reads from the raw sequencing data of the TEd-Seq, effectively removing the background noise and non-specific sequences. Additionally, we filter low-quality and sequencing artefacts.

- #### Generate longer contigs
High-quality paired-end reads are merged to generate longer contigs, facilitating accurate identification of transposon insertion-genome split sites. 

 - #### Mapping to reference genome and detection of unique/split (soft-clipped) alignments
Assembled contigs are aligned to the reference genome in local alignement mode to enable split-read mapping. The resulting alignments are then filtered to retain only unique and split (soft-clipped) alignments wherein the split reagion is atleast ~40nt in length. 

- #### Mapping to targeted TE extermity
In parallel, the assembled contigs are also aligned to the extremity sequence of the targeted TE family using local alignement mode and TE-specific alignments are retrieved.

- #### Tagging Alignments
In this crucial step, the alignments to the reference genome and targeted TEs are collated to preserve split contigs with shared read homology between the target TE and reference genomic sites. The TE read-tagged alignments are then used for calling new insertion sites.

- #### Calling non-reference somatic and heritable transposition events
Novel TE insertions are classified as either _de novo_ somatic or heritable (segregating) transposition events based on read depth at the insertion sites. All new insertion events marked by a read coverage of ≤5 are regarded as somatic events while TE insertions with a threshold of ≥30 reads were classified heritable transpositions. The identified putative insertions are filtered to remove ambiguous events.

Further details of the pipeline can be found in Ambreen et al. 

## Requirements
deNOVOEnrich is executed through a Bash script and is designed for command-line use on Unix-based systems.

- Linux 86X64 Systems
- cutadapt (https://cutadapt.readthedocs.io/en/stable/)
- fastp (https://github.com/OpenGene/fastp)
- seqtk (https://github.com/lh3/seqtk)
- trimmomatic (http://www.usadellab.org/cms/?page=trimmomatic)
- PEAR (https://github.com/tseemann/PEAR)
- bowtie2 (https://bowtie-bio.sourceforge.net/bowtie2/index.shtml)
- picard (https://broadinstitute.github.io/picard/)
- samtools (https://www.htslib.org/)
- bedtools (https://bedtools.readthedocs.io/en/latest/index.html)
- readtagger (https://pypi.org/project/readtagger/)

Ensure that all required dependencies are properly installed in the working environment to avoid errors during run. 

## Input files
- Reference genome [fasta format]
- Reference TE files
  - TE extremity sequences. Two fasta files containing the TE extremity region (from th primer site to 5'LTR extremity) and its reverse complement.  
  - Known native/reference loci of the TE family [BED file]. The bed file listing genomic coordinates of the TE's native loci in the reference genome. These are used to filter out false positives, which often cluster near the native loci due to sequencing artifacts.
- Mono-nucleotide genomic sites [BED file]. The bed file containing genomic coordiates of mononucleotide repeats, can be generated by tools such as MIcroSAtellite (MISA) identification tool (https://webblast.ipk-gatersleben.de/misa/) and further filtering to retain mononucleotide repeats.








